py/builtinevex.c

```diff
@@ -136,17 +136,18 @@ STATIC mp_obj_t eval_exec_helper(size_t n_args, const mp_obj_t *args, mp_parse_i
     }
     #endif
 
-    // Extract the source code.
-    mp_buffer_info_t bufinfo;
-    mp_get_buffer_raise(args[0], &bufinfo, MP_BUFFER_READ);
 
     // create the lexer
     // MP_PARSE_SINGLE_INPUT is used to indicate a file input
     mp_lexer_t *lex;
     if (MICROPY_PY_BUILTINS_EXECFILE && parse_input_kind == MP_PARSE_SINGLE_INPUT) {
-        lex = mp_lexer_new_from_file(bufinfo.buf);
+        lex = mp_lexer_new_from_file(mp_obj_str_get_qstr(args[0]));
```

Comment:
Couldn't the fix simply be to use `mp_obj_str_get_str` here?

Reply:
Yes but now you have two copies of the string. One made here, another made in `mp_reader_new_file`.

Also one day we might want to not make str and/or qstr null terminated, so I'm always hesitant to make this dependency worse.

Reply:
Well, `mp_obj_str_get_str` doesn't create anything, it just returns a pointer to the already-null-terminated data.  So using that function would be a very simple fix for this problem, and not allocate any additional RAM.  Then the larger qstr change is an optimisation.